### ISSUE TITLE
Exact solutions

### ISSUE THREAD

**Issue by dswatson** [mayer79/kernelshap#22]
I think we can adapt the existing code for exact solutions to extend the approach beyond $p = 5$ while actually speeding things up. It might go something like this.

1. We don't need any data to compute `A`, which only depends on `p`. So a simple utils function will do the trick. (Fun fact, the formula for off-diagonal values in Covert & Lee's Appx. A is actually wrong. You can verify this by comparing it with brute force results as in your existing `crossprod(Z) / n_Z` when `Z` is exact.)
```r
# Here's the proposed utils function
off_diag <- function(p) {
  S <- 1:(p - 1)
  probs <- (p - 1) / (choose(p, S) * S * (p - S))
  probs <- probs / sum(probs)
  c_pr <- (0.5 * S^2 - 0.5 * S) / choose(p, 2)
  sum(probs * c_pr)
}

# Then A is just given by
if (exact) {
  A <- diag(0.5, nrow = p)
  A[A == 0] <- off_diag(p)
  ...
}
```
2. The tougher part is computing `b` without any extra rows beyond the $2^p - 2$ required for complete enumeration of `Z`-vectors. The trick is just to use those weights that are currently computed as part of `create_exact_Z`, but deliver them along with the  matrix `Z`. This can be done outside the `kernelshap_one` loop to avoid unnecessary computation.
```r
# Another utils function, this one to replace create_exact_Z
complete_Z <- function(p) {
  # Create Z
  z <- do.call(expand.grid, replicate(p, 0:1, simplify = FALSE))
  z <- z[2:(nrow(z) - 1), ]
  z <- as.matrix(z)
  dimnames(z) <- NULL
  rs <- rowSums(z)
  # Compute weights
  S <- 1:(p - 1)
  probs <- (p - 1) / (choose(p, S) * S * (p - S))
  probs <- probs / sum(probs)
  w <- probs / as.vector(table(rs))
  w <- w[rs]
  out <- list('Z' = z, 'w' = w)
  return(out)
}

# Then we can resume the code above
if (exact) {
  A <- diag(0.5, nrow = p)
  A[A == 0] <- off_diag(p)
  cz <- complete_Z(p)
  Z <- cz$Z
  w <- cz$w
  b <- crossprod(Z, w * (vz - v0_ext))
}
```
Of course, the value functions depend on the data, so `b` will need to be computed within the `kernelshap_one` loop. Everything else can be done outside though, and requires only $2^p - 2$ calls to `pred_fun` per sample, which is less than the current `create_exact_Z` implementation. Also allows for arbitrary $p$ if people are willing to wait for exact results. 

Just a thought, anyway! Happy to create a pull request from the current branch if you're interested.

**Comment by mayer79** [comment]
Fantastic thoughts!

I did the following:

1. Submitted the current status to CRAN. I wanted to be quite fast with version 0.2.0 in order to have as few people being affected by the new interface.  Submit does not mean "accepted", so it is not yet on CRAN and maybe we can even bring this cool idea into v0.2.0.
2. Added a new util function `kernel_weights()`.
3. Slightly optimized the complete_Z function:
```
complete_Z <- function(p) {
  if (p < 2L) {
    stop("p must be at least two")
  }
  Z <- as.matrix(do.call(expand.grid, replicate(p, 0:1, simplify = FALSE)))
  dimnames(Z) <- NULL
  Z <- Z[2:(nrow(Z) - 1L), , drop = FALSE]
  rs <- rowSums(Z)
  w <- kernel_weight(p) / tabulate(rs, nbins = p - 1L)
  list(Z = Z, w = w[rs])
}
```
4. What is "k" in off_diag?
5. Currently, kernelshap_one() makes only one call (per iteration) to the predict function. This happens in get_vz(). There, m * nrow(bg_X) rows are sent to pred_fun(), which quickly reaches 1 mio if p around 10-12. Simple solution: we currently limit exact calculations to p = 10 and do not have to change much. More sophisticated solution: we chuck Z in blocks of 1000 and call get_vz() for each block. This is a bit tougher as we need to change the logic or creating replicated bg_X and X.

**Comment by dswatson** [comment]
Sounds good re: (1)-(3). Apologies on (4), `k` was a typo, I meant `S` (revised now). 

As for (5), I think it's probably fine to max out around p = 10. I would recommend a warning instead of an error though. Different users may have different needs/hardware. The block solution is a good idea for exact results on super big jobs, but personally I would probably just opt for the sample version in that case? Do you know how they handle this in SHAP? 

I'll create a new pull request from the current main branch so we're all lined up.

**Comment by mayer79** [comment]
Sketch of kernelshap_one(): It separates exact calculations altogether from sampling calculations. It is not tested though...

```
# Kernel SHAP algorithm for a single row x with paired sampling
kernelshap_one <- function(object, X, bg_X, pred_fun, bg_w, v0, v1, 
                           paired, m, exact, tol, max_iter, ...) {
  p <- ncol(X)
  n_Z <- m * (1L + paired)
  v0_ext <- v0[rep(1L, n_Z), , drop = FALSE]                      #  (n_Z x K)

  if (exact) {
    A <- matrix(off_diag(p), nrow = p, ncol = p)
    diag(A) <- 0.5
    cz <- complete_Z(p)
    vz <- get_vz(                                                 #  (n_Z x K)
      X = X, bg = bg_X, Z = cz$Z, object = object, pred_fun = pred_fun, w = bg_w, ...
    )
    # Note: cz$w is replicated along cols of (vz - v0_ext)
    b <- crossprod(cz$Z, cz$w * (vz - v0_ext))
    beta <- solver(A, b, v1, v0)
    
    return(list(beta = beta, sigma = 0 * beta, n_iter = 1L, converged = TRUE))
  }
  
  # Now the sampling version
  est_m = list()
  converged <- FALSE
  n_iter <- 0L
  Asum <- matrix(0, nrow = p, ncol = p)                           #  (p x p)
  bsum <- matrix(0, nrow = p, ncol = ncol(v0))                    #  (p x K)

  while(!isTRUE(converged) && n_iter < max_iter) {
    n_iter <- n_iter + 1L
    
    # Create Z matrix of dimension ->                             #  (n_Z x p)
    Z <- sample_Z(m = m, p = p)
    if (paired) {
      Z <- rbind(Z, 1 - Z)
    }
    
    # Calling get_vz() is expensive                               #  (n_Z x K)
    vz <- get_vz(
      X = X, bg = bg_X, Z = Z, object = object, pred_fun = pred_fun, w = bg_w, ...
    )
    
    # Least-squares with constraint that beta_1 + ... + beta_p = v_1 - v_0. 
    # The additional constraint beta_0 = v_0 is dealt via offset
    Atemp <- crossprod(Z) / n_Z                                   #  (p x p)
    btemp <- crossprod(Z, (vz - v0_ext)) / n_Z                    #  (p x K)
    Asum <- Asum + Atemp                                          #  (p x p)
    bsum <- bsum + btemp                                          #  (p x K)
    est_m[[n_iter]] <- R <- solver(Atemp, btemp, v1, v0)          #  (p x K)
    
    # Covariance calculation would fail in the first iteration
    if (n_iter >= 2L) {
      beta_n <- solver(Asum / n_iter, bsum / n_iter, v1, v0)      #  (p x K)
      sigma_n <- get_sigma(est_m, iter = n_iter)                  #  (p x K)
      converged <- all(conv_crit(sigma_n, beta_n) < tol)
    }
  }
  list(beta = beta_n, sigma = sigma_n, n_iter = n_iter, converged = converged)
}
```

**Comment by dswatson** [comment]
This looks great! Feel free to close the issue once the changes are pushed. Probably not much point in that pull request now :) 

I'm not entirely sure I follow how SHAP chooses between sampling and exact regimes. What does it mean to check if there is enough "weight" for a given k? Are some values of $k \in \lbrace 2, ..., p-1 \rbrace$ evaluated exactly and others only approximately? 

Another SHAP functionality that could be worth exploring is regularization penalties, e.g. $l_1$. I'll give that a think, maybe open another issue if I can come up with any suggestions.

**Comment by mayer79** [comment]
Now as we understand how to deal with weights, , we could even move into the direction of Python's "shap", improving it by the iterative logic of Covert. The exact case would simply pop as a side result.

This is my interpretation how "shap" works (sorry for the lengthy code):

```
make_Z <- function(p, m) {
  p_half <- ceiling((p - 1) / 2)
  uneven <- (p - 1) %% 2
  K <- seq_len(p_half)
  
  # How many z are we allowed to create per k?
  available_m <- kernel_weights(p)[K] * m * 2
  if (uneven) {
    available_m[p_half] <- available_m[p_half] / 2
  }

  # How many z do exist per k?
  all_z_per_k <- choose(p, K) * 2

  Z <- vector("list", length = p - 1L)
  for (k in K) {
    # Exact
    if (all_z_per_k[k] <= available_m[k]) {
      Z[[k]] <- RcppAlgos::permuteGeneral(0:1, freq = c(p - k, k))
      unused <- available_m[k] - nrow(Z[[k]])
      available_m[k] <- nrow(Z[[k]])
      if (k < p_half && unused > 0L) {
        s <- (k + 1L):p_half
        available_m[s] <- available_m[s] * (1 + unused / sum(available_m[s]))
      }
    } else{
      # Sampling (could also do in paired way)
      Z[[k]] <- t(
        replicate(
          available_m[k], {out <- numeric(p); out[sample(1:p, k)] <- 1; out}
        )
      )
    }
    if (k < p_half || !uneven) {
      Z[[p - k]] <- 1L - Z[[k]]
    }
  }
  Z <- do.call(rbind, Z)
  
  rs <- rowSums(Z)
  w <- kernel_weights(p) / tabulate(rs, nbins = p - 1L)
  list(Z = Z, w = w[rs])
}
```

We would calculate 

- Atemp as crossprod(cz$Z, cz$w * cz$Z) (which corresponds to your solution if m <= 2^p-2)
- btemp as proposed: b <- crossprod(cz$Z, cz$w * (vz - v0_ext)) 

If m <= 2^p-2, we would iterate just once.

This would depreciate arguments "exact" and "paired_sampling". Probably, we would need to do a bit less preallocations as the size of Z might be slightly off.

Side notes:

- I would not dig too much into L1, it violates the nice theory behind kernelSHAP
- Our tabulate(rowsums...) is simply `choose(p, S)`
- version 0.2.0 was accepted by CRAN, so we have 3-4 weeks time until the next update will be possible.

**Comment by mayer79** [comment]
Actually, we can keep the exact case separate and think of improving the current sampling approach by Scott Lundberg's logic.

**Comment by dswatson** [comment]
This looks great. I haven't gone through the code in close detail but I get the gist. I think there's probably some merit in having an `exact` flag just so people are sure. (One small comment – I think the `for` loop could be implemented more efficiently with `sapply`?) 

As for L1 penalties, I'm not sure I agree. For high-dimensional problems, it can be decidedly uninformative to learn that all of your, say, 500 predictors contribute to a given prediction "a little bit". What we really want in such cases is a selection mechanism, some way to say "I get that all variables matter, but if I had to explain predictions in terms of just the top 10, how would I do it?" Both LIME and SHAP offer some version of this. I'm not entirely sure how it would work in this setting though. I don't think we can just make a call to `glmnet`, for instance, and all accompanying theory for standard errors would definitely go out the window.

In any event, congrats on the CRAN acceptance! Sounds like we've got some time to mull over any further amendments. If I come up with anything for the L1 penalty I'll file a new issue.
